File: sdk/memory-core/typescript/src/v3/ops-client.ts
/**
 * v3 OpsClient — Knowledge Service ops plane over the shared V3HttpTransport.
 *
 * - `llmBindingSet / llmBindingStatus / llmBindingList` →
 *   POST /v3/internal/llm-binding/{set,status,list} (per-instance LLM routing;
 *   api_key is never echoed back; list needs no service-id).
 *   First set for an instance requires api_key server-side — omit it only
 *   when updating an already-bound instance (server keeps the stored key).
 * - `autoSyncStatus` → GET /v3/auto-sync/status (scheduler snapshot + config).
 * - `autoSyncTrigger` → POST /v3/auto-sync/trigger (fire-and-forget scan;
 *   `{triggered:false}` when disabled server-side).
 *
 * Auth: x-tdai-service-id always sent; Bearer only if apiKey given.
 */
import { ParamError } from "../errors.js";
import { V3HttpTransport } from "./http.js";
import type { Transport } from "../client.js";
import { stripUndefined } from "./request-helpers.js";
import type {
  AutoSyncStatusResult, AutoSyncTriggerResult, LlmBindingListResult,
  LlmBindingSetRequest, LlmBindingSetResult, LlmBindingStatus, OpsClientConfig,
} from "./ops-types.js";
const V3BIND = "/v3/internal/llm-binding";
const V3SYNC = "/v3/auto-sync";
function need(name: string, v: string | undefined): string {
  if (!v || !v.trim()) throw new ParamError("OpsClient requires non-empty " + name);
  return v;
}
function checkSet(p: LlmBindingSetRequest): void {
  if (p.mode !== "proxy" && p.mode !== "byo") throw new ParamError("llmBindingSet mode must be 'proxy' or 'byo'");
  if (p.mode === "proxy" && !p.proxy_base_url) throw new ParamError("llmBindingSet proxy mode requires proxy_base_url");
  if (p.mode === "byo" && !p.base_url) throw new ParamError("llmBindingSet byo mode requires base_url");
}
export class OpsClient {
  private readonly http: Transport;
  constructor(config: OpsClientConfig);
  constructor(transport: Transport);
  constructor(configOrTransport: OpsClientConfig | Transport) {
    if ("post" in configOrTransport) {
      this.http = configOrTransport;
      if (typeof this.http.get !== "function") {
        throw new ParamError("OpsClient transport requires get() (auto-sync/status is GET)");
      }
      return;
    }
    const cfg = configOrTransport;
    need("serviceId", cfg.serviceId);
    if (!cfg.endpoint || !cfg.endpoint.trim()) throw new ParamError("OpsClient requires non-empty endpoint");
    this.http = new V3HttpTransport({
      endpoint: cfg.endpoint, apiKey: cfg.apiKey, serviceId: cfg.serviceId,
      timeout: cfg.timeout, rejectUnauthorized: cfg.rejectUnauthorized,
    });
  }
  /** Upsert per-instance LLM routing. Idempotent — re-post overwrites. */
  llmBindingSet(p: LlmBindingSetRequest): Promise<LlmBindingSetResult> {
    checkSet(p);
    return this.http.post(V3BIND + "/set", stripUndefined({
      mode: p.mode, api_key: p.api_key, proxy_base_url: p.proxy_base_url, base_url: p.base_url, enabled: p.enabled,
    }));
  }
  /** Read binding status (never contains api_key). */
  llmBindingStatus(): Promise<LlmBindingStatus> {
    return this.http.post(V3BIND + "/status", {});
  }
  /** List all bindings (no service-id needed server-side). */
  llmBindingList(): Promise<LlmBindingListResult> {
    return this.http.post(V3BIND + "/list", {});
  }
  /** Scheduler snapshot + config. Only GET in the KS API. */
  autoSyncStatus(): Promise<AutoSyncStatusResult> {
    return this.http.get!(V3SYNC + "/status");
  }
  /** Fire-and-forget scan trigger. `{triggered:false}` when disabled. */
  autoSyncTrigger(): Promise<AutoSyncTriggerResult> {
    return this.http.post(V3SYNC + "/trigger", {});
  }
}

File: sdk/memory-core/python/tencentdb_agent_memory/v3/ops_client.py
"""TencentDB Agent Memory v3 Ops SDK — Knowledge Service ops plane (shared ``HttpStub``).

- LLM binding (``POST /v3/internal/llm-binding/{set,status,list}``):
  per-instance LLM routing; ``api_key`` is never echoed back; ``list``
  needs no service-id. First set for an instance requires ``api_key``
  server-side — omit it only when updating an already-bound instance.
- Auto-sync (``GET /v3/auto-sync/status``, ``POST /v3/auto-sync/trigger``):
  scheduler snapshot + manual fire-and-forget scan trigger.

Auth: ``x-tdai-service-id`` always sent; ``Authorization: Bearer`` only
when ``api_key`` is given (KS default needs none).
"""

from __future__ import annotations

from typing import Any, Dict, Optional

from .._v3_http import AsyncHttpStub, HttpStub
from ..errors import ParamError
from ._common import ks_stub, need, strip_none

_V3BIND = "/v3/internal/llm-binding"
_V3SYNC = "/v3/auto-sync"


def _check_set(p: Dict[str, Any]) -> None:
    mode = p.get("mode")
    if mode not in ("proxy", "byo"):
        raise ParamError("llm_binding_set mode must be 'proxy' or 'byo'")
    if mode == "proxy" and not p.get("proxy_base_url"):
        raise ParamError("llm_binding_set proxy mode requires proxy_base_url")
    if mode == "byo" and not p.get("base_url"):
        raise ParamError("llm_binding_set byo mode requires base_url")


class _OpsMethodsMixin:
    _stub: Any

    def _llm_binding_set(self, p: Dict[str, Any]) -> Dict[str, Any]:
        _check_set(p)
        return self._stub.post(f"{_V3BIND}/set", strip_none({
            "mode": p.get("mode"), "api_key": p.get("api_key"),
            "proxy_base_url": p.get("proxy_base_url"), "base_url": p.get("base_url"),
            "enabled": p.get("enabled"),
        }))

    def _llm_binding_status(self) -> Dict[str, Any]:
        return self._stub.post(f"{_V3BIND}/status", {})

    def _llm_binding_list(self) -> Dict[str, Any]:
        return self._stub.post(f"{_V3BIND}/list", {})

    def _auto_sync_status(self) -> Dict[str, Any]:
        return self._stub.get(f"{_V3SYNC}/status")

    def _auto_sync_trigger(self) -> Dict[str, Any]:
        return self._stub.post(f"{_V3SYNC}/trigger", {})


class OpsClient(_OpsMethodsMixin):
    """Sync KS ops client (llm-binding + auto-sync)."""

    def __init__(self, endpoint: str = "", service_id: Optional[str] = None, *, api_key: str = "", timeout: float = 30, verify: bool = True, stub: Optional[Any] = None) -> None:
        self._stub = stub if stub is not None else ks_stub(endpoint, service_id, api_key=api_key, timeout=timeout, verify=verify)

    def llm_binding_set(self, p: Dict[str, Any]) -> Dict[str, Any]:
        return self._llm_binding_set(p)

    def llm_binding_status(self) -> Dict[str, Any]:
        return self._llm_binding_status()

    def llm_binding_list(self) -> Dict[str, Any]:
        return self._llm_binding_list()

    def auto_sync_status(self) -> Dict[str, Any]:
        return self._auto_sync_status()

    def auto_sync_trigger(self) -> Dict[str, Any]:
        return self._auto_sync_trigger()

    def close(self) -> None:
        c = getattr(self._stub, "close", None)
        if callable(c):
            c()

    def __enter__(self) -> "OpsClient":
        return self

    def __exit__(self, *exc: Any) -> None:
        self.close()


class AsyncOpsClient(_OpsMethodsMixin):
    """Async KS ops client — same surface as OpsClient."""

    def __init__(self, endpoint: str = "", service_id: Optional[str] = None, *, api_key: str = "", timeout: float = 30, verify: bool = True, stub: Optional[Any] = None) -> None:
        self._stub = stub if stub is not None else ks_stub(endpoint, service_id, api_key=api_key, timeout=timeout, verify=verify, async_=True)

    async def llm_binding_set(self, p: Dict[str, Any]) -> Dict[str, Any]:
        return await self._llm_binding_set(p)

    async def llm_binding_status(self) -> Dict[str, Any]:
        return await self._llm_binding_status()

    async def llm_binding_list(self) -> Dict[str, Any]:
        return await self._llm_binding_list()

    async def auto_sync_status(self) -> Dict[str, Any]:
        return await self._auto_sync_status()

    async def auto_sync_trigger(self) -> Dict[str, Any]:
        return await self._auto_sync_trigger()
